src/Microsoft.AspNetCore.Mvc.TagHelpers/CacheTagHelperBase.cs

```diff
@@ -0,0 +1,305 @@
+                    .Append(nameof(VaryByUser))
+                    .Append(CacheKeyTokenSeparator)
+                    .Append(ViewContext.HttpContext.User?.Identity?.Name);
+            }
+
+            // The key is typically too long to be useful, so we use a cryptographic hash
+            // as the actual key (better randomization and key distribution, so small vary
+            // values will generate dramatically different keys).
+            using (var sha = SHA256.Create())
+            {
+                var contentBytes = Encoding.UTF8.GetBytes(builder.ToString());
+                var hashedBytes = sha.ComputeHash(contentBytes);
+                return Convert.ToBase64String(hashedBytes);
+            }
+        }
+
+        protected static void AddStringCollectionKey(
+            StringBuilder builder,
+            string keyName,
+            string value,
+            IDictionary<string, StringValues> sourceCollection)
+        {
+            if (!string.IsNullOrEmpty(value))
+            {
+                // keyName(param1=value1|param2=value2)
+                builder
+                    .Append(CacheKeyTokenSeparator)
+                    .Append(keyName)
+                    .Append("(");
+
+                var values = Tokenize(value);
+
+                // Perf: Avoid allocating enumerator
+                for (var i = 0; i < values.Count; i++)
+                {
+                    var item = values[i];
+                    builder
+                        .Append(item)
+                        .Append(CacheKeyTokenSeparator)
+                        .Append(sourceCollection[item])
+                        .Append(CacheKeyTokenSeparator);
+                }
+
+                if (values.Count > 0)
+                {
+                    // Remove the trailing separator
+                    builder.Length -= CacheKeyTokenSeparator.Length;
+                }
+
+                builder.Append(")");
+            }
+        }
+
+        protected static void AddStringCollectionKey<TSourceCollection>(
+            StringBuilder builder,
+            string keyName,
+            string value,
+            TSourceCollection sourceCollection,
+            Func<TSourceCollection, string, StringValues> accessor)
+        {
+            if (!string.IsNullOrEmpty(value))
+            {
+                // keyName(param1=value1|param2=value2)
+                builder
+                    .Append(CacheKeyTokenSeparator)
+                    .Append(keyName)
+                    .Append("(");
+
+                var values = Tokenize(value);
+
+                // Perf: Avoid allocating enumerator
+                for (var i = 0; i < values.Count; i++)
+                {
+                    var item = values[i];
+
+                    builder
+                        .Append(item)
+                        .Append(CacheKeyTokenSeparator)
+                        .Append(accessor(sourceCollection, item))
```

Comment:
Do you need the `accessor`? Unless i'm missing something it looks like it always evaluates to `sourceCollection[item]`